src/Stratis.Bitcoin.Features.Consensus/CoinviewPrefetcher.cs

```diff
@@ -0,0 +1,112 @@
+    public class CoinviewPrefetcher : IDisposable
+    {
+        /// <summary>
+        /// How many blocks ahead prefetching will look.
+        /// When header at height X is dequeued block with header at height <c>X + Lookahead</c> will be prefetched in case block data is downloaded.
+        /// </summary>
+        private const int Lookahead = 5;
+
+        /// <summary>Queue of headers that were added when block associated with such header was fully validated.</summary>
+        private readonly AsyncQueue<ChainedHeader> headersQueue;
+
+        private readonly ICoinView coinview;
+
+        private readonly CoinviewHelper coinviewHelper;
+
+        private readonly ConcurrentChain chain;
+
+        private readonly ILogger logger;
+
+        public CoinviewPrefetcher(ICoinView coinview, ConcurrentChain chain, ILoggerFactory loggerFactory)
+        {
+            this.coinview = coinview;
+            this.chain = chain;
+
+            this.headersQueue = new AsyncQueue<ChainedHeader>(this.OnHeaderEnqueuedAsync);
+            this.coinviewHelper = new CoinviewHelper();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Prefetches UTXOs for a block and some blocks with higher height.
+        /// Prefetching is done on the background.
+        /// </summary>
+        /// <param name="header">Header of a block that is about to be partially validated and requires prefetching.</param>
+        private async Task OnHeaderEnqueuedAsync(ChainedHeader header, CancellationToken cancellation)
+        {
+            ChainedHeader currentHeader = header;
+
+            // Go Lookahead blocks ahead of current header and get block for prefetching.
+            // There might be several blocks at height of header.Height + Lookahead but
+            // only first one will be prefetched since prefetching is in place mostly to
+            // speed up IBD.
+            for (int i = 0; i < Lookahead; i++)
+            {
+                if (currentHeader.Next.Count == 0)
+                {
+                    this.logger.LogTrace("(-)[NO_HEADERS]");
+                    return;
+                }
+
+                currentHeader = currentHeader.Next.FirstOrDefault();
+
+                if (currentHeader == null)
+                {
+                    this.logger.LogTrace("(-)[NO_NEXT_HEADER]");
+                    return;
+                }
+            }
+
+            Block block = currentHeader.Block;
+
+            if (block == null)
+            {
+                this.logger.LogTrace("(-)[NO_BLOCK_DATA]");
+                return;
+            }
+
+            bool enforceBIP30 = DeploymentFlags.EnforceBIP30ForBlock(currentHeader);
+            uint256[] idsToFetch = this.coinviewHelper.GetIdsToFetch(block, enforceBIP30);
+
+            bool farFromTip = currentHeader.Height > this.chain.Tip.Height + 2;
```

Comment:
Why is this necessary?
currentHeader is always tip +5

Reply:
Because we are working in background thread and we might advance consensus too fast so by the time we want to prefetch it's already not needed.

Reply:
hmm good point, its probably not worth adding a dependency on CH for that though.
Can you use chain instead?

Reply:
Didn't quite understand what you mean

Reply:
Don't use `ConcurrentChain`